### PR TITLE
Add just and pre-commit

### LINKED ISSUE #187: [Quality] Pre-commit hooks
## Problem
We have GitHub actions that expect a certain code style be adhered to, but this isn't enforced as a pre-commit hook, making it annoying for developers to have to fix these only after the action reports the problem.

## Description
Using the `pre-commit` package we can:

0. Add a pre-commit hook that will lint all our files
1. Replace the GitHub action script with a simple `pre-commit run --all-files`

## Implementation

- [x] 🙋 I would be interested in implementing this feature.

### LINKED ISSUE #193: [Quality] Use black and isort
## Current Situation
We do not use an automatic formatter.

## Suggested Improvement
Let's use an automatic formatter and import sorter so we don't have to think about it anymore. `black` is the standard opinionated Python formatter, and isort is the standard import sorter. Both are very safe and battle tested.

We will exclude the archive directory from them though.

## Benefit
We won't have to think about formatting any longer 🙂 

## Implementation

- [x] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #193 by @sarayourfriend 
Fixes #187 by @sarayourfriend 

## Description
Adds `pre-commit` using the same configuration as the `openverse-catalog` with some modifications to ignore rules in certain files that were problematic.

Also adds a `justfile` to make docker stuff easier.

## Technical details
I had to add some `noqa` comments for long strings (I wish flake8 would ignore strings...)

I also added `.flake8` that mirrors the pre-commit configuration for flake8 so that editor integration will work

## Tests
Checkout the branch and run `just install` then `pipenv run pre-commit run --all-files` and ensure that there are no errors reported.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by dhruvkb** [comment]
@sarayourfriend we should remove the 'Style' job from the 'Automated tests' workflow as the new 'Linting' job supersedes it (the Flake8 check in 'Linting' already includes the `pycodestyle` check from 'Style').

**Comment by sarayourfriend** [comment]
Oops, I meant to remove it when I added the linting job 😅 It's gone now, thanks!

File: justfile
Comment (AetherUnbound) [inline justfile:37]: It might be possible to have a single `logs` command, with optional service:
```suggestion
logs service="":
    docker-compose {{ DOCKER_FILE }} logs -f {{ service }}
```

If nothing is supplied, we get `logsall`, but otherwise individual services can be chosen.

File: justfile
Comment (AetherUnbound) [inline justfile:4]: It doesn't seem like `SERVICE_NAME` is used - is there an intention for it in the future?

**Review by AetherUnbound — APPROVED** [review]
Looks great! I only have some additional thoughts on the `justfile`

File: justfile
Comment (sarayourfriend) [reply justfile:37]: Good idea!

File: justfile
Comment (sarayourfriend) [reply justfile]: Whoops, I think it's just left over. It's removed now :)

**Review by krysal — COMMENTED** [review]
It looks pretty good! I just leave some suggestions and questions to evaluate.

File: .pre-commit-config.yaml
Comment (sarayourfriend) [reply .pre-commit-config.yaml:49]: It can't 😞 `pre-commit` runs each repo in its own virtualenv and exists ?? somewhere ?? that doesn't include the root files from our repository, so it doesn't have access to our `.flake8` file 😢

File: README.md
Comment (sarayourfriend) [reply README.md:30]: Yup! I'll add a link to `just`

File: ingestion_server/ingestion_server/es_mapping.py
Comment (sarayourfriend) [reply ingestion_server/ingestion_server/es_mapping.py:28]: We can exclude the file if you prefer! I think black tries to print lines on as few as possible until it has to introduce a line break? So it compresses lines like this.

Comment (krysal) [reply ingestion_server/ingestion_server/es_mapping.py]: Yes, this is a special case, thanks!

File: .pre-commit-config.yaml
Comment (krysal) [reply .pre-commit-config.yaml:49]: I see, the multi-project issue, then don't mind my question. Thanks for the clarification.

File: README.md
Comment (krysal) [inline README.md:129]: I got this by trying to run this command:

```sh
error: Justfile does not contain recipe `testlocal`
```

Comment (sarayourfriend) [reply README.md:129]: Oh hmm I removed it for some reason. Let me see.

Do you know why these tests must run in a "pipenv subshell"?

Comment (krysal) [reply README.md:129]: Personally, I run the tests inside the container, we can leave only those instructions for tests.

```shell
docker-compose exec web bash ./test/run_test.sh
```

Comment (sarayourfriend) [reply README.md:129]: I wonder if it's ever necessary to run it outside of the container 🤔

Comment (dhruvkb) [reply README.md:129]: Sometimes, I run the code outside the container when debugging it line by line. Running the tests outside the containers helps in that scenario. I'm open to a different process though if I'm the only one using it like that.

Comment (sarayourfriend) [reply README.md:129]: Yeah that seems like a reasonable use case. Can't `PyCharm` attach to the in-container interpreter though? I never tried to use the GUI debugger for that though, I always relied on `pdb` 😞 the GUI is so much nicer when it's available. I'll think about some alternative approaches to this problem.

Comment (krysal) [reply README.md:129]: Yes, Pycharm can take the interpreter of a container, it's quite flexible in this aspect. However, I can see the value in the local approach in not being tied up to this particular IDE, neither to the use of `just` if that is what is blocking here.

Comment (dhruvkb) [reply README.md:129]: @krysal from what I had read in the docs, PyCharm can work with a Docker environment only if it created the container. I wasn't able to get it to connect with the existing `web` container orchestrated by Docker Compose 😢.

**Comment by dhruvkb** [comment]
@sarayourfriend this might be a silly suggestion but I was playing around with pre-commit in the current setup and came across something that might be a good compromise.

Let's say we installed `pre-commit` in both the existing `Pipfile`s while keeping the `.pre-commit-config.yaml` in the repo root. To set up the hooks we go into either of the repos and invoke `pipenv run pre-commit install`.

The hook created records the virtualenv that was active when `install` was invoked so there's no need to always activate the virtualenv when committing. All pre-commit hooks manage their own environment so there's no need for installing `black` or `flake8` on any project `Pipfile`.

What do you think are the cons of this approach?

**Comment by sarayourfriend** [comment]
> Apart from having to list pre-commit in two Pipfiles, what do you think are the cons of this approach?

The real cons are for editor integration, but we're already sacrificing the best of editor integration (dependency knowledge) in the current set up, so I think what you propose is better.

We can add it to the `openverse-api/Pipfile`?

**Comment by dhruvkb** [comment]
Let's use `just install` to install from the `openverse-api` directory. That's the main project in the repo (it shouldn't matter but still)

File: README.md
Comment (sarayourfriend) [reply README.md:129]: Hmm, I don't think that's true Dhruv, the docs might be incorrect. At my previous job we were able to get it to connect to the docker container just fine and it was running using docker compose... but maybe things have changed since then.

**Comment by sarayourfriend** [comment]
@dhruvkb I did as you suggested and it works great 🙂

File: README.md
Comment (sarayourfriend) [reply README.md:129]: In any case, after making the changes in the latest commits we're able to add back the `testlocal` script 🎉

File: .pre-commit-config.yaml
Comment (dhruvkb) [inline .pre-commit-config.yaml:42]: Just to match the official way their org is named.
```suggestion
  - repo: https://gitlab.com/PyCQA/flake8
```

**Review by dhruvkb — COMMENTED** [review]
To finalise the changes, it'll be good to squash the commits in such a way that
- all formatting changes are squashed into one commit to [avoid ruining Git blame](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html#avoiding-ruining-git-blame)
- all config changes are squashed into one commit (to make reviewing the PR easier)

**Comment by dhruvkb** [comment]
I wonder why the tests are still failing. I'll pull this locally to review so will also look into that then.

**Comment by sarayourfriend** [comment]
@dhruvkb Sure, I can do that. Alternatively I could also merge using the `squash and merge` strategy so it will all be under one commit. Do you have a strong preference either way?

**Comment by dhruvkb** [comment]
I have no hard preferences, whatever is convenient for you is cool.

File: .github/workflows/integration-tests.yml
Comment (dhruvkb) [inline .github/workflows/integration-tests.yml:23]: Should specify [`working-directory`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsrun) instead.

**Review by dhruvkb — COMMENTED** [review]
I'll review this tomorrow.

File: .github/workflows/integration-tests.yml
Comment (sarayourfriend) [reply .github/workflows/integration-tests.yml:23]: Oh cool, I didn't know about that!

File: .flake8
Comment (dhruvkb) [inline .flake8:3]: It would be nice to include the names of the ignored rules. It'll be helpful to know what checks we're skipping.

File: .pre-commit-config.yaml
Comment (dhruvkb) [reply .pre-commit-config.yaml:42]: Their [docs](https://flake8.pycqa.org/en/latest/user/using-hooks.html) show a GitHub URL. Let's use that.
```suggestion
  - repo: https://github.com/PyCQA/flake8
```

File: .pre-commit-config.yaml
Comment (dhruvkb) [inline .pre-commit-config.yaml:30]: `name` should be used consistently for all hooks or none. Naming one hook is a bit weird.

File: .pre-commit-config.yaml
Comment (dhruvkb) [inline .pre-commit-config.yaml:49]: I think these args won't be necessary as the hook will automatically use the config from `.flake8`.

File: README.md
Comment (dhruvkb) [inline README.md:30]: Add `just` to the prerequisites section above. `just` commands are so much simpler that it's practically endorsable as a the recommended approach.

File: README.md
Comment (dhruvkb) [reply README.md:129]: The `testlocal` script works perfectly. So does the regular `test` one.

File: justfile
Comment (dhruvkb) [inline justfile:24]: The unquoted URL for cURL causes issues with shells like `zsh`. Not when running with `just` but when running as a command directly. 

I'd recommend an optional small change.
```suggestion
    curl "localhost:8000/v1/images/?q=honey"
```

File: justfile
Comment (dhruvkb) [inline justfile:3]: Good to have in case we split to multiple Compose files. Thanks for future-proofing this!

**Review by dhruvkb — APPROVED** [review]
A linter, formatter and Git hook trio is a much needed addition to the repo! Thanks for doing this. 

And with passing tests, I have a lot of confidence in this PR. Apart from all the breaking changes this'll cause in all other PRs (especially the big one #194), LGTM.

**Comment by dhruvkb** [comment]
@sarayourfriend is there a possibility of holding the merge for this PR and merging other open PRs first? It'll be very hard to update them with `master` after this is merged, while this one will be easier to update if we just took the configs and re-ran the hooks.

File: .pre-commit-config.yaml
Comment (dhruvkb) [inline .pre-commit-config.yaml:45]: I missed the indentation problem here...

File: .pre-commit-config.yaml
Comment (dhruvkb) [inline .pre-commit-config.yaml:54]: ...and here.

**Review by dhruvkb — COMMENTED** [review]
There are some indentation inconsistencies in the pre-commit config. Is it weird that `pre-commit` didn't catch this?

File: .pre-commit-config.yaml
Comment (sarayourfriend) [reply .pre-commit-config.yaml:49]: I won't, unfortunately. `pre-commit` runs hooks inside a virtualenv that doesn't include the root of your project, so it doesn't know any of your configuration files 😢

**Comment by sarayourfriend** [comment]
> is there a possibility of holding the merge for this PR and merging other open PRs first

Sure! I suppose it'd be easier to just let the repo settle down at some point and then redo this PR from scratch... the merge conflicts will be difficult to resolve either way (usually these types of PRs are best applied quickly before too much work is done in the meantime, but this one has taken longer 😞) so it'll be easy to just copy the configuration from this PR and then run the formatting.

I guess can you let me know when you think this PR should be merged? Otherwise we'll be waiting forever for work to settle 🙂 At some point we just have to accept the complications.

File: .pre-commit-config.yaml
Comment (sarayourfriend) [reply .pre-commit-config.yaml]: Removed

File: README.md
Comment (sarayourfriend) [reply README.md:30]: Done

**Comment by sarayourfriend** [comment]
Oh also, indentation is fixed! Thanks for pointing it out.

File: .pre-commit-config.yaml
Comment (dhruvkb) [reply .pre-commit-config.yaml]: I tried this locally and it seemed to work fine. Are you sure about it not working?

Comment (dhruvkb) [reply .pre-commit-config.yaml]: @sarayourfriend for the hook, is it possible to provide a config path instead of defining the rules twice in two different locations? If not I think just leaving a note here should be enough that mentions the rules also being in the other file.

Comment (sarayourfriend) [reply .pre-commit-config.yaml]: I guess I'm not sure 😅 I'd tried it in the past and it didn't work and I found some SO answers about it to confirm the issue, but if you've got it working I'll try it again!

Comment (sarayourfriend) [reply .pre-commit-config.yaml]: Ah it worked! I removed the duplicated config. Thanks for testing it out!

**Comment by zackkrida** [comment]
How about we merge this after these are merged/closed?

CleanShot 2021-09-19 at 19 28 14@2x

**Comment by dhruvkb** [comment]
#202 will need to be reworked anyway, we can merge that later. But #194 is too big to redo 😢.

**Comment by sarayourfriend** [comment]
Closing in favor of https://github.com/WordPress/openverse-api/pull/224

**Pull request closed without merging**
